path_checkout_sets.go

```diff
@@ -0,0 +1,413 @@
+package openldap
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/go-secure-stdlib/strutil"
+	"github.com/hashicorp/vault/sdk/framework"
+	"github.com/hashicorp/vault/sdk/helper/locksutil"
+	"github.com/hashicorp/vault/sdk/logical"
+)
+
+const libraryPrefix = "library/"
+
+type librarySet struct {
+	ServiceAccountNames       []string      `json:"service_account_names"`
+	TTL                       time.Duration `json:"ttl"`
+	MaxTTL                    time.Duration `json:"max_ttl"`
+	DisableCheckInEnforcement bool          `json:"disable_check_in_enforcement"`
+}
+
+// Validate ensures that a set meets our code assumptions that TTLs are set in
+// a way that makes sense, and that there's at least one service account.
+func (l *librarySet) Validate() error {
+	if len(l.ServiceAccountNames) < 1 {
+		return fmt.Errorf(`at least one service account must be configured`)
+	}
+	if l.MaxTTL > 0 {
+		if l.MaxTTL < l.TTL {
+			return fmt.Errorf(`max_ttl (%d seconds) may not be less than ttl (%d seconds)`, l.MaxTTL, l.TTL)
+		}
+	}
+	return nil
+}
+
+func (b *backend) pathListSets() []*framework.Path {
+	return []*framework.Path{
+		{
+			Pattern: libraryPrefix + "?$",
+			Operations: map[logical.Operation]framework.OperationHandler{
+				logical.ListOperation: &framework.PathOperation{
+					Callback: b.setListOperation,
+				},
+			},
+			HelpSynopsis:    pathListSetsHelpSyn,
+			HelpDescription: pathListSetsHelpDesc,
+		},
+	}
+}
+
+func (b *backend) setListOperation(ctx context.Context, req *logical.Request, _ *framework.FieldData) (*logical.Response, error) {
+	keys, err := req.Storage.List(ctx, libraryPrefix)
```

Comment:
It could be inconsistent, but I can't think of any significant consequences. I haven't seen synchronization involved in list operations like this. It will return the list from storage at that point in time.